lib/adds.h

```diff
@@ -0,0 +1,76 @@
+// SPDX-FileCopyrightText: 2023, Alejandro Colomar <alx@kernel.org>
+// SPDX-License-Identifier: BSD-3-Clause
+
+
+#ifndef SHADOW_INCLUDE_LIB_ADDS_H_
+#define SHADOW_INCLUDE_LIB_ADDS_H_
+
+
+#include <config.h>
+
+#include <errno.h>
+#include <limits.h>
+#include <stdlib.h>
+
+#include "sizeof.h"
+
+
+inline long addsl(long a, long b);
+inline long addsl3(long a, long b, long c);
+
+inline int cmpl(const void *p1, const void *p2);
+
+
+inline long
+addsl(long a, long b)
+{
+	if (a > 0 && b > LONG_MAX - a) {
+		errno = EOVERFLOW;
+		return LONG_MAX;
+	}
+	if (a < 0 && b < LONG_MIN - a) {
+		errno = EOVERFLOW;
+		return LONG_MIN;
+	}
+	return a + b;
+}
+
+
+inline long
+addsl3(long a, long b, long c)
+{
+	int   e;
+	long  sum;
+	long  n[3];
+
+	n[0] = a;
+	n[1] = b;
+	n[2] = c;
+
+	e = errno;
+	qsort(n, NITEMS(n), sizeof(n[0]), cmpl);
```

Comment:
What's the advantage of sorting?

Reply:
The order of addition matters: consider `LONG_MAX + 1 + LONG_MIN`.

```c
LONG_MAX + 1 + LONG_MIN
```

In the above, LONG_MAX + 1 already saturates, so we should return early and report saturation if we didn't sort.  If we sort, we get the right sum:

```
LONG_MAX + LONG_MIN + 1 == 0
```

The sorting makes it so that we add the more positive with the more negative number, so that we don't overflow intermediate additions (unless it's really necessary).  It is the optimum algorithm for avoiding overflowing the type in intermediate operations.

The addsN(...) general algorithm for adding N signed integers would be to add the most positive and the most negative numbers (so, [0] and [N-1] after the sort), and then pass that sum with the rest of the array to an addsN-1() function, recursively handling the operation, sorting N-2 times.

Reply:
I've written a non-recursive generalized version of the algorithm for arbitrary N, just for satisfying my curiosity.

I've only tested it for N==3, but it should work for any N, I think.

It's a bit more complex, so I'm not adding it to shadow, unless you ask for it.

```c
// Copyright 2023, Alejandro Colomar <alx@kernel.org>
// SPDX-License-Identifier: GPL-3.0-or-later

#define addslN(a, b, ...)                                                     \
({                                                                            \
	int   e;                                                              \
	long  n[] = {a, b, __VA_ARGS__};                                      \
                                                                              \
	e = errno;                                                            \
	for (size_t i = NITEMS(n); i > 2; i--) {                              \
		qsort(n, i, sizeof(n[0]), cmpl);                              \
                                                                              \
		errno = 0;                                                    \
		n[0] = addsl(n[0], n[i-1]);                                   \
		if (errno == EOVERFLOW) {                                     \
			n[1] = n[0];                                          \
			e = errno;                                            \
			break;                                                \
		}                                                             \
	}                                                                     \
	errno = e;                                                            \
                                                                              \
	addsl(n[0], n[1]);                                                    \
})
```

Reply:
addslN looks good to me.

Reply:
Thanks!  :-)
Happy new year!